<a href="https://colab.research.google.com/github/hannahrfong/CSC713M/blob/main/Investigatory_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **CSC713M Investigatory Project**

# Importing Libraries

In [88]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
#import scipy.stats as sps

In [89]:
fie_df = pd.read_csv('https://raw.githubusercontent.com/hannahrfong/CSC713M/main/fie.csv', na_values=["NA"], keep_default_na=False)

# Dataset Description

## Brief Description

The Family Income and Expenditure (FIE) dataset is derived from the 2015 Family Income and Expenditure Survey (FIES) conducted by the Philippine Statistics Authority (PSA). The dataset provides data on the income and expenditure of Filipino households which can be used to formulate governmental policies and programs to elevate the socio-economic conditions of Filipinos.

## Dataset File Structure

The dataset contains 41,544 samples and 60 features. Each sample represents one Filipino household.

In [90]:
fie_df.shape

(41544, 60)

The list of features in the dataset and their corresponding data types can be generated by running the code below.

In [91]:
fie_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41544 entries, 0 to 41543
Data columns (total 60 columns):
 #   Column                                         Non-Null Count  Dtype 
---  ------                                         --------------  ----- 
 0   Total Household Income                         41544 non-null  int64 
 1   Region                                         41544 non-null  object
 2   Total Food Expenditure                         41544 non-null  int64 
 3   Main Source of Income                          41544 non-null  object
 4   Agricultural Household indicator               41544 non-null  int64 
 5   Bread and Cereals Expenditure                  41544 non-null  int64 
 6   Total Rice Expenditure                         41544 non-null  int64 
 7   Meat Expenditure                               41544 non-null  int64 
 8   Total Fish and  marine products Expenditure    41544 non-null  int64 
 9   Fruit Expenditure                              41544 non-null

# Data Cleaning

The FIE dataset will undergo data cleaning to identify noisy, incomplete, and inconsistent data that can degrade the performance of machine learning models to be used with the dataset.

Specifically, the following aspects of the dataset will be checked:

*   Multiple Representations
*   Incorrect Datatypes
*   Default Values
*   Missing Data
*   Duplicate Data
*   Inconsistent Format









## Multiple Representations

The `unique` function is used to check if there exist different representations in the categorical features of the dataset.

In [92]:
for column in fie_df.select_dtypes(include="object"):
    print("'{}' unique values:\n".format(column), fie_df[column].unique())
    print("")

'Region' unique values:
 ['CAR' 'Caraga' 'VI - Western Visayas' 'V - Bicol Region' ' ARMM'
 'III - Central Luzon' 'II - Cagayan Valley' 'IVA - CALABARZON'
 'VII - Central Visayas' 'X - Northern Mindanao' 'XI - Davao Region'
 'VIII - Eastern Visayas' 'I - Ilocos Region' 'NCR' 'IVB - MIMAROPA'
 'XII - SOCCSKSARGEN' 'IX - Zasmboanga Peninsula']

'Main Source of Income' unique values:
 ['Wage/Salaries' 'Other sources of Income' 'Enterpreneurial Activities']

'Household Head Sex' unique values:
 ['Female' 'Male']

'Household Head Marital Status' unique values:
 ['Single' 'Married' 'Widowed' 'Divorced/Separated' 'Annulled' 'Unknown']

'Household Head Highest Grade Completed' unique values:
 ['Teacher Training and Education Sciences Programs'
 'Transport Services Programs' 'Grade 3' 'Elementary Graduate'
 'Second Year High School' 'Third Year High School'
 'Business and Administration Programs' 'First Year College'
 'High School Graduate'
 'Other Programs in Education at the Third Level, Firs

Upon inspection, it can be confirmed that there are no multiple representations in the dataset.

## Incorrect Datatype

The `info` function is used to validate the data types of the values in the dataset.

In [93]:
fie_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41544 entries, 0 to 41543
Data columns (total 60 columns):
 #   Column                                         Non-Null Count  Dtype 
---  ------                                         --------------  ----- 
 0   Total Household Income                         41544 non-null  int64 
 1   Region                                         41544 non-null  object
 2   Total Food Expenditure                         41544 non-null  int64 
 3   Main Source of Income                          41544 non-null  object
 4   Agricultural Household indicator               41544 non-null  int64 
 5   Bread and Cereals Expenditure                  41544 non-null  int64 
 6   Total Rice Expenditure                         41544 non-null  int64 
 7   Meat Expenditure                               41544 non-null  int64 
 8   Total Fish and  marine products Expenditure    41544 non-null  int64 
 9   Fruit Expenditure                              41544 non-null

Given that the categorical features are an object type and the numerical features have an integer (`int64`) type, no incorrect datatypes were found amongst all the features.

## Default Values

The `unique` function is used to identify default values in the dataset and to assess if these are appropriate in the context of the project.

In [94]:
for column in fie_df:
    print("'{}' unique values:\n".format(column), fie_df[column].unique())
    print("")

'Total Household Income' unique values:
 [480332 198235  82785 ... 133171 129500 128598]

'Region' unique values:
 ['CAR' 'Caraga' 'VI - Western Visayas' 'V - Bicol Region' ' ARMM'
 'III - Central Luzon' 'II - Cagayan Valley' 'IVA - CALABARZON'
 'VII - Central Visayas' 'X - Northern Mindanao' 'XI - Davao Region'
 'VIII - Eastern Visayas' 'I - Ilocos Region' 'NCR' 'IVB - MIMAROPA'
 'XII - SOCCSKSARGEN' 'IX - Zasmboanga Peninsula']

'Total Food Expenditure' unique values:
 [117848  67766  61609 ...  31157  81416  78195]

'Main Source of Income' unique values:
 ['Wage/Salaries' 'Other sources of Income' 'Enterpreneurial Activities']

'Agricultural Household indicator' unique values:
 [0 1 2]

'Bread and Cereals Expenditure' unique values:
 [42140 17329 34182 ... 19693 28563  2691]

'Total Rice Expenditure' unique values:
 [38300 13008 32001 ... 15918 23457  1273]

'Meat Expenditure' unique values:
 [24676 17434  7783 ...  6905  3181  2359]

'Total Fish and  marine products Expenditure' un

The result shows that the `Agricultural Household indicator` feature has 3 unique values (i.e. `0`, `1`, `2`). However, the metadata of the dataset specifies that there are only 2 possible values for this feature:
*   1 - Agricultural Household
*   2 - Non-agricultural Household


In [95]:
fie_df.loc[fie_df['Agricultural Household indicator'] == 0].shape

(28106, 60)

There are 28,106 observations with the undefined `0` value for the `Agricultural Household indicator` feature, which is more than half (67.65%) of the dataset. Therefore, these observations cannot just be deleted.

Instead, the `Agricultural Household indicator` feature will be removed to handle the ambiguous data and prevent it from impairing the performance of the machine learning models to be used.

In [96]:
fie_df = fie_df.drop(['Agricultural Household indicator'], axis=1)
fie_df.shape

(41544, 59)

## Missing Data

The `isnull` and `any` functions are used to check for missing values in the dataset.

In [97]:
fie_df.isnull().any()

Total Household Income                           False
Region                                           False
Total Food Expenditure                           False
Main Source of Income                            False
Bread and Cereals Expenditure                    False
Total Rice Expenditure                           False
Meat Expenditure                                 False
Total Fish and  marine products Expenditure      False
Fruit Expenditure                                False
Vegetables Expenditure                           False
Restaurant and hotels Expenditure                False
Alcoholic Beverages Expenditure                  False
Tobacco Expenditure                              False
Clothing, Footwear and Other Wear Expenditure    False
Housing and water Expenditure                    False
Imputed House Rental Value                       False
Medical Care Expenditure                         False
Transportation Expenditure                       False
Communicat

In [98]:
nan_variables = fie_df.columns[fie_df.isnull().any()].tolist()
print(nan_variables)

['Household Head Occupation', 'Household Head Class of Worker']


In [99]:
fie_df[nan_variables].isnull().sum()

Household Head Occupation         7536
Household Head Class of Worker    7536
dtype: int64

There are 2 features with missing values:

1. Household Head Occupation
2. Household Head Class of Worker

Let us check if these null values are missing at random or not by counterchecking with the `Household Head Job or Business Indicator` feature that indicates whether a household head is employed or not.



In [100]:
job_variables = nan_variables
job_variables.append('Household Head Job or Business Indicator')

fie_df[job_variables].loc[(fie_df['Household Head Occupation'].isnull()) & (fie_df['Household Head Class of Worker'].isnull()) & (fie_df['Household Head Job or Business Indicator'] == 'No Job/Business')]

Household Head Occupation Household Head Class of Worker   
8                           NaN                            NaN  \
13                          NaN                            NaN   
14                          NaN                            NaN   
15                          NaN                            NaN   
26                          NaN                            NaN   
...                         ...                            ...   
41520                       NaN                            NaN   
41529                       NaN                            NaN   
41533                       NaN                            NaN   
41535                       NaN                            NaN   
41543                       NaN                            NaN   

      Household Head Job or Business Indicator  
8                              No Job/Business  
13                             No Job/Business  
14                             No Job/Business  
15                             No Job/Business  
26                             No Job/Business  
...                                        ...  
41520                          No Job/Business  
41529                          No Job/Business  
41533                          No Job/Business  
41535                          No Job/Business  
41543                          No Job/Business  

[7536 rows x 3 columns]

The executed code above returns observations that have a `NaN` value for `Household Head Occupation`, a `Nan` value for `Household Head of Class of Worker`, and a `No Job/Business` value for `Household Head Job or Business Indicator`. The result contains 7,536 observations which is the same number of observations that have null values for the 2 `nan_variables` (i.e. `Household Head Occupation` and `Household Head of Class of Worker`).

Based on this, it can be deduced that the missing values are not missing at random. Instead, they indicate the unemployment of household heads.

The null values will be handled by replacing them with an `Unemployed` value to properly label the data.

In [101]:
fie_df = fie_df.fillna('Unemployed')
nan_variables = fie_df.columns[fie_df.isnull().any()].tolist()
print(nan_variables)

[]


After running the code above, there are no more missing values in the dataset.

## Duplicate Data

The `drop_duplicates` function is used to delete duplicate data in the dataset.

In [102]:
fie_df.drop_duplicates()

Total Household Income              Region  Total Food Expenditure   
0                      480332                 CAR                  117848  \
1                      198235                 CAR                   67766   
2                       82785                 CAR                   61609   
3                      107589                 CAR                   78189   
4                      189322                 CAR                   94625   
...                       ...                 ...                     ...   
41539                  119773  XII - SOCCSKSARGEN                   44875   
41540                  137320  XII - SOCCSKSARGEN                   31157   
41541                  133171  XII - SOCCSKSARGEN                   45882   
41542                  129500  XII - SOCCSKSARGEN                   81416   
41543                  128598  XII - SOCCSKSARGEN                   78195   

            Main Source of Income  Bread and Cereals Expenditure   
0                   Wage/Salaries                          42140  \
1                   Wage/Salaries                          17329   
2                   Wage/Salaries                          34182   
3                   Wage/Salaries                          34030   
4                   Wage/Salaries                          34820   
...                           ...                            ...   
41539  Enterpreneurial Activities                          23675   
41540  Enterpreneurial Activities                           2691   
41541  Enterpreneurial Activities                          28646   
41542  Enterpreneurial Activities                          29996   
41543  Enterpreneurial Activities                          43485   

       Total Rice Expenditure  Meat Expenditure   
0                       38300             24676  \
1                       13008             17434   
2                       32001              7783   
3                       28659             10914   
4                       30167             18391   
...                       ...               ...   
41539                   21542              1476   
41540                    1273              1886   
41541                   27339               480   
41542                   26655              2359   
41543                   41205              1985   

       Total Fish and  marine products Expenditure  Fruit Expenditure   
0                                            16806               3325  \
1                                            11073               2035   
2                                             2590               1730   
3                                            10812                690   
4                                            11309               1395   
...                                            ...                ...   
41539                                         6120               1632   
41540                                         4386               1840   
41541                                         4796               1232   
41542                                        17730               2923   
41543                                         7735               2062   

       Vegetables Expenditure  ...  Number of Refrigerator/Freezer   
0                       13460  ...                               1  \
1                        7833  ...                               0   
2                        3795  ...                               0   
3                        7887  ...                               0   
4                       11260  ...                               1   
...                       ...  ...                             ...   
41539                    3882  ...                               0   
41540                    3110  ...                               0   
41541                    3025  ...                               0   
41542                    7951  ...                               0   
41543

There are 41,544 observations returned after running the `drop_duplicates` function, which is the same as the number of observations in the original dataset. Therefore, there are no duplicates in the dataset.

## Inconsistent Format

The Philippine regions have an inconsistent format, particularly for regions IX – Zamboanga Peninsula, XIII – Caraga, and ARMM. Region IX contains an misspelling of the region, Region XIII does not have its roman numeral as part of its label, and ARMM contains a whitespace character before the abbreviation.

In [103]:
for column in fie_df:
    print("'{}' unique values:\n".format(column), fie_df[column].unique())
    print("")

'Total Household Income' unique values:
 [480332 198235  82785 ... 133171 129500 128598]

'Region' unique values:
 ['CAR' 'Caraga' 'VI - Western Visayas' 'V - Bicol Region' ' ARMM'
 'III - Central Luzon' 'II - Cagayan Valley' 'IVA - CALABARZON'
 'VII - Central Visayas' 'X - Northern Mindanao' 'XI - Davao Region'
 'VIII - Eastern Visayas' 'I - Ilocos Region' 'NCR' 'IVB - MIMAROPA'
 'XII - SOCCSKSARGEN' 'IX - Zasmboanga Peninsula']

'Total Food Expenditure' unique values:
 [117848  67766  61609 ...  31157  81416  78195]

'Main Source of Income' unique values:
 ['Wage/Salaries' 'Other sources of Income' 'Enterpreneurial Activities']

'Bread and Cereals Expenditure' unique values:
 [42140 17329 34182 ... 19693 28563  2691]

'Total Rice Expenditure' unique values:
 [38300 13008 32001 ... 15918 23457  1273]

'Meat Expenditure' unique values:
 [24676 17434  7783 ...  6905  3181  2359]

'Total Fish and  marine products Expenditure' unique values:
 [16806 11073  2590 ... 10623 12496 17730]

'Fr

In [104]:
regions = {
    'IX - Zasmboanga Peninsula': 'IX - Zamboanga Peninsula',
    ' ARMM': 'ARMM',
    'Caraga': 'XIII - Caraga'
}

fie_df['Region'] = fie_df['Region'].map(regions).fillna(fie_df['Region'])

With this, we map these errors to the correct labels by using the `map()` function and the `regions` dictionary.

The `Type of Walls` feature has a subset of samples that contain a `NOt applicable` value. To make this consistent with other columns that have a `Not Applicable` value, we replace all current instances with the new value.

In [105]:
fie_df['Type of Walls'] = fie_df['Type of Walls'].replace('NOt applicable', 'Not Applicable')

The `Total Income from Entrepreneurial Acitivites` feature misspells the word *Activities*.

In [106]:
fie_df.rename(columns={'Total Income from Entrepreneurial Acitivites': 'Total Income from Entrepreneurial Activites'})

Total Household Income              Region  Total Food Expenditure   
0                      480332                 CAR                  117848  \
1                      198235                 CAR                   67766   
2                       82785                 CAR                   61609   
3                      107589                 CAR                   78189   
4                      189322                 CAR                   94625   
...                       ...                 ...                     ...   
41539                  119773  XII - SOCCSKSARGEN                   44875   
41540                  137320  XII - SOCCSKSARGEN                   31157   
41541                  133171  XII - SOCCSKSARGEN                   45882   
41542                  129500  XII - SOCCSKSARGEN                   81416   
41543                  128598  XII - SOCCSKSARGEN                   78195   

            Main Source of Income  Bread and Cereals Expenditure   
0                   Wage/Salaries                          42140  \
1                   Wage/Salaries                          17329   
2                   Wage/Salaries                          34182   
3                   Wage/Salaries                          34030   
4                   Wage/Salaries                          34820   
...                           ...                            ...   
41539  Enterpreneurial Activities                          23675   
41540  Enterpreneurial Activities                           2691   
41541  Enterpreneurial Activities                          28646   
41542  Enterpreneurial Activities                          29996   
41543  Enterpreneurial Activities                          43485   

       Total Rice Expenditure  Meat Expenditure   
0                       38300             24676  \
1                       13008             17434   
2                       32001              7783   
3                       28659             10914   
4                       30167             18391   
...                       ...               ...   
41539                   21542              1476   
41540                    1273              1886   
41541                   27339               480   
41542                   26655              2359   
41543                   41205              1985   

       Total Fish and  marine products Expenditure  Fruit Expenditure   
0                                            16806               3325  \
1                                            11073               2035   
2                                             2590               1730   
3                                            10812                690   
4                                            11309               1395   
...                                            ...                ...   
41539                                         6120               1632   
41540                                         4386               1840   
41541                                         4796               1232   
41542                                        17730               2923   
41543                                         7735               2062   

       Vegetables Expenditure  ...  Number of Refrigerator/Freezer   
0                       13460  ...                               1  \
1                        7833  ...                               0   
2                        3795  ...                               0   
3                        7887  ...                               0   
4                       11260  ...                               1   
...                       ...  ...                             ...   
41539                    3882  ...                               0   
41540                    3110  ...                               0   
41541                    3025  ...                               0   
41542                    7951  ...                               0   
41543

# Data Preprocessing

## Binning

### Total Household Income

The observations will be binned into the different income classes as defined by the Philippine Institute for Development Studies (Domingo, 2020). The binning will be based on the monthly household income, which is obtained by dividing the `Total Household Income` by 6 since it reflects the total household income from January to June 2015 (Philippine statistics Authority, 2017).  

In [107]:
monthly_income = fie_df['Total Household Income'] / 6
bins = [0, 10957, 21914, 43828, 76669, 131484, 219140, monthly_income.max() + 1]
labels = ["Poor", "Low-income", "Lower middle", "Middle", "Upper middle", "Upper middle but not rich", "Rich"]
fie_df['Income Class'] = pd.cut(monthly_income, bins=bins, labels=labels, right=False)
fie_df['Income Class'] = fie_df['Income Class'].astype('object')
fie_df['Income Class'] 

0        Upper middle
1        Lower middle
2          Low-income
3          Low-income
4        Lower middle
             ...     
41539      Low-income
41540    Lower middle
41541    Lower middle
41542      Low-income
41543      Low-income
Name: Income Class, Length: 41544, dtype: object

### Household Head Highest Grade Completed

Based on the “Household Head Highest Grade Completed”, there are multiple labels that dictate a specific educational attainment that a person has completed. To simplify the labels, we bin these values into five categories: `No Grade Completed`, `Preschool`, `Primary Level`, `Secondary Level`, and `Tertiary Level`.

In [108]:
grade = {
    'Preschool': 'Preschool',
    'No Grade Completed': 'No Grade Completed',
    'Grade 1': 'Primary Level',
    'Grade 2': 'Primary Level',
    'Grade 3': 'Primary Level',
    'Grade 4': 'Primary Level',
    'Grade 5': 'Primary Level',
    'Grade 6': 'Primary Level',
    'Elementary Graduate': 'Primary Level',
    'First Year High School': 'Secondary Level',
    'Second Year High School': 'Secondary Level',
    'Third Year High School': 'Secondary Level',
    'High School Graduate': 'Secondary Level',
    'Post Baccalaureate': 'Post Baccalaureate'
}

fie_df['Household Head Highest Grade Completed'] = fie_df['Household Head Highest Grade Completed'].map(grade)
fie_df['Household Head Highest Grade Completed'] = fie_df['Household Head Highest Grade Completed'].replace(np.nan, 'Tertiary Level')

# Exploratory Data Analysis

The attributes of the FIES dataset can be generally classified into 3 categories: demographics, socio-economic featueres, and expenditures. 

In line with these types of data, the researchers formulated the following EDA questions:

1. How is the dataset distributed based on demographics?
2. What are the characteristics of the different income classes?
3. How does expenditure differ across the different income classes?

## EDA Question #1

The first EDA question aims to determine how the data is distributed based on demographics.

Specifically, we aim to answer the following specific questions:
1. How many observations are there per region?
2. What is the distribution of the age of household heads?
3. What is the ratio of male to female household heads?
4. What is the ratio of male to female per class of worker? 
5. What is the average total household income per region?
6. What is the employment rate of household heads per region? // is it better to compare between regions?

### 1. Number of Observations Per Region

In [ ]:
region_counts = fie_df['Region'].value_counts()

plt.bar(region_counts.index, region_counts.values)

plt.xlabel('Region')
plt.ylabel('Count')
plt.title('Number of Observations per Region')

plt.xticks(rotation=75)
plt.show()

### 2. Distribution of Household Head Age

In [ ]:
data = fie_df['Household Head Age']
mean_age = data.mean()

plt.hist(data, bins=50, edgecolor='w')

x_axis_labels = range(0, 101, 10)
plt.xticks(x_axis_labels)
plt.xlabel('Household Head Age')
plt.ylabel('Count')

plt.axvline(mean_age, color='red', linestyle='dashed', linewidth=2, label=f'Mean Age = {mean_age}')

plt.legend()
plt.show()

### 3. Ratio of Male to Female Household Heads

In [ ]:
hhs_counts = fie_df['Household Head Sex'].value_counts()
plt.figure(figsize=(10,6))
hhs_counts.plot(kind='bar', rot=0)
plt.xlabel('Household Head Sex')
plt.ylabel('Count')
plt.title('Household Head Sex')
plt.show()

### 4. Ratio of Male to Female Per Class of Worker

In [ ]:
grouped = fie_df.groupby(['Household Head Class of Worker', 'Household Head Sex']).size().unstack(fill_value=0)
sex_totals = grouped.sum(axis=1)
grouped_ratio = grouped.divide(sex_totals, axis=0)
grouped_ratio

ax = grouped_ratio.plot(kind='bar', stacked=True, figsize=(10, 6))
ax.set_xlabel('Household Head Class of Worker')
ax.set_ylabel('Ratio of Household Head Sex')
ax.set_title('Ratio of Sex Per Class of Worker')
plt. legend(title='Household Head Sex', loc='upper right')

plt.show()

### 5. Average Total Household Income Per Region

In [ ]:
income_mean = fie_df.groupby('Region').agg({'Total Household Income': ['mean']})
income_mean.sort_values(by=('Total Household Income', 'mean'), ascending=False)

In [ ]:
income_mean = income_mean.sort_values(by=('Total Household Income', 'mean'), ascending=False) 
income_mean.plot(kind='bar', legend=None)
plt.xlabel('Region')
plt.ylabel('Average Total Household Income')
plt.title('Average Total Household Income Per Region')

### 6. Household Head Employment Rate Per Region - todo: change to employment rate of families?

In [ ]:
grouped = fie_df.groupby(['Region', 'Household Head Job or Business Indicator']).size().unstack(fill_value=0)
employment_totals = grouped.sum(axis=1)
grouped_ratio = grouped.divide(employment_totals, axis=0)
grouped_ratio

ax = grouped_ratio.plot(kind='bar', stacked=True, figsize=(10, 6))
ax.set_xlabel('Region')
ax.set_ylabel('Ratio of Employment')
ax.set_title('Ratio of Employment Per Region')
plt. legend(title='Employment', loc='upper right')

## EDA Question #2

The second EDA question aims to identify the characteristics of the different income classes.

Specifically, we aim to answer the following questions:
1. What is the distribution of households per income class?
2. What is the distribution of total household income of each income class?
3. What is the proportion of educational attainment of household heads per income class?
4. What is the proportion of main source of income per income class?
5. What is the proportion of class of worker of household heads per income class?
6. What are the occupations of the household heads of the top 5 poorest and richest households?

### 1. Distribution of Households Per Income Class

In [ ]:
value_counts = fie_df['Income Class'].value_counts()
custom_order = ["Poor", "Low-income", "Lower middle", "Middle", "Upper middle", "Upper middle but not rich", "Rich"]
sorted_value_counts = value_counts.reindex(custom_order)
sorted_value_counts

In [ ]:
plt.bar(sorted_value_counts.index, sorted_value_counts.values)
plt.xlabel('Count')
plt.ylabel('Income Class')
plt.title('Income Distribution')
plt.xticks(rotation=90)  
plt.show()

### 2. Distribution of Total Household Income per Income Class

In [ ]:
income_df = fie_df[['Total Household Income', 'Income Class']]

# Define the order of income classes
income_class_order = ["Poor", "Low-income", "Lower middle", "Middle", "Upper middle", "Upper middle but not rich", "Rich"]

# Reorder the 'Income Class' column based on the defined order
income_df['Income Class'] = pd.Categorical(income_df['Income Class'], categories=income_class_order, ordered=True)

income_df.boxplot('Total Household Income', by='Income Class', figsize=(15,10))
plt.title('Total Household Income by Income Class')
plt.show()

### 3. Proportion of Household Head Educational Attainment By Income Class

In [ ]:
custom_order = ["Poor", "Low-income", "Lower middle", "Middle", "Upper middle", "Upper middle but not rich", "Rich"]

grouped = fie_df.groupby(['Income Class', 'Household Head Highest Grade Completed']).size().unstack(fill_value=0)
grouped = grouped.reindex(custom_order)
educ_totals = grouped.sum(axis=1)
grouped_ratio = grouped.divide(educ_totals, axis=0)
ax = grouped_ratio.plot(kind='bar', stacked=True, figsize=(10, 6))
ax.set_xlabel('Income Class')
ax.set_ylabel('Ratio')
ax.set_title('Proportion of Household Head Educational Attainment by Income Class')
plt.legend(title='Educational Attainment', loc='upper right')

plt.show()

### 4. Proportion of Main Source of Income By Income Class

In [ ]:
custom_order = ["Poor", "Low-income", "Lower middle", "Middle", "Upper middle", "Upper middle but not rich", "Rich"]

grouped = fie_df.groupby(['Income Class', 'Main Source of Income']).size().unstack(fill_value=0)
grouped = grouped.reindex(custom_order)
class_totals = grouped.sum(axis=1)
grouped_ratio = grouped.divide(class_totals, axis=0)

ax = grouped_ratio.plot(kind='bar', stacked=True, figsize=(10, 6))
ax.set_xlabel('Income Class')
ax.set_ylabel('Ratio of Main Source of Income')
ax.set_title('Proportion of Main Source of Income By Income Class')
plt.legend(title='Main Source of Income', loc='upper right')

plt.show()

### 5. Proportion of Household Head Class of Worker Per Income Class

In [ ]:
custom_order = ["Poor", "Low-income", "Lower middle", "Middle", "Upper middle", "Upper middle but not rich", "Rich"]

grouped = fie_df.groupby(['Income Class', 'Household Head Class of Worker']).size().unstack(fill_value=0)
grouped = grouped.reindex(custom_order)
class_totals = grouped.sum(axis=1)
grouped_ratio = grouped.divide(class_totals, axis=0)

ax = grouped_ratio.plot(kind='bar', stacked=True, figsize=(10, 6))
ax.set_xlabel('Income Class')
ax.set_ylabel('Ratio of Household Head Class of Worker')
ax.set_title('Proportion of Household Head Class of Worker By Income Class')
plt.legend(title='Household Head Class of Worker', loc='upper right')

plt.show()

### 6. Household Head Occupations of Top 5 Poorest and Richest Households

In [ ]:
top_5_poorest = fie_df.nsmallest(5, 'Total Household Income')

# Get the Top 5 richest households based on 'Household Income'
top_5_richest = fie_df.nlargest(5, 'Total Household Income')

# Extract occupations of household heads for the Top 5 Poorest and Richest households
occupations_poorest = top_5_poorest['Household Head Occupation']
occupations_richest = top_5_richest['Household Head Occupation']

# Display the results
print("Top 5 Poorest Household Head Occupations:")
print(occupations_poorest)

print("\nTop 5 Richest Household Head Occupations:")
print(occupations_richest)

## EDA Question #3

The third EDA question aims to distinguish the expenditure pattern among the different income classes by answering the following questions:
1. What is the relationship between income and expense?
2. What are the expenditure proportions of each income class across different expense categories?

### 1. Relationship Between Income and Expense

In [ ]:
expenses = [
    'Total Food Expenditure',
    'Restaurant and hotels Expenditure',
    'Alcoholic Beverages Expenditure',
    'Tobacco Expenditure',
    'Clothing, Footwear and Other Wear Expenditure',
    'Housing and water Expenditure',
    'Imputed House Rental Value',
    'Medical Care Expenditure',
    'Transportation Expenditure',
    'Communication Expenditure',
    'Education Expenditure',
    'Miscellaneous Goods and Services Expenditure',
    'Special Occasions Expenditure',
    'Crop Farming and Gardening expenses'
]

total_expense = fie_df[expenses].sum(axis=1)
plt.scatter(fie_df['Total Household Income'], total_expense, color='b', alpha=0.5)  # Use alpha to adjust point transparency


# Labeling the axes and title
plt.xlabel('Total Household Income')
plt.ylabel('Total Expenditure')
plt.title('Total Expenditure by Total Household Income')

# Show grid
plt.grid(True)

# Display the plot
plt.show()

### 2. Expenditure Proportions Per Income Class

In [ ]:
expenses = [
    'Total Food Expenditure',
    'Restaurant and hotels Expenditure',
    'Alcoholic Beverages Expenditure',
    'Tobacco Expenditure',
    'Clothing, Footwear and Other Wear Expenditure',
    'Housing and water Expenditure',
    'Imputed House Rental Value',
    'Medical Care Expenditure',
    'Transportation Expenditure',
    'Communication Expenditure',
    'Education Expenditure',
    'Miscellaneous Goods and Services Expenditure',
    'Special Occasions Expenditure',
    'Crop Farming and Gardening expenses'
]

expense_scatter = fie_df.groupby("Income Class")[expenses].mean().round(2).sort_values(by=expenses, ascending=False)
custom_order = ["Poor", "Low-income", "Lower middle", "Middle", "Upper middle", "Upper middle but not rich", "Rich"]
expense_scatter = expense_scatter.reindex(custom_order)

ax = expense_scatter.plot(kind='bar', rot=70, stacked=True, figsize=(10,6))
ax.set_xlabel('Income Class')
ax.set_ylabel('Total of the Average Expense Per Category')
ax.set_title('Expenses Per Income Class')
plt.legend(title='Expense Categories', loc='upper right')

plt.show()

# Data Transformation

## Dropping of Irrelevant Variables

In [109]:
fie_df = fie_df.drop(['Household Head Job or Business Indicator', 'Total Household Income'], axis=1)
fie_df.shape

(41544, 58)

## One Hot Encoding of Nominal Variables

### Preprocessing

In [110]:
# Others
fie_df['Toilet Facilities'] = fie_df['Toilet Facilities'].replace('Others', 'Others - Toilet Facilities')
fie_df['Main Source of Water Supply'] = fie_df['Main Source of Water Supply'].replace('Others', 'Others - Water Supply')

# Unemployed
fie_df['Household Head Occupation'] = fie_df['Household Head Occupation'].replace('Unemployed', 'Unemployed - HHO')
fie_df['Household Head Class of Worker'] = fie_df['Household Head Class of Worker'].replace('Unemployed', 'Unemployed - HHCW')

# Not Applicable
fie_df['Type of Roof'] = fie_df['Type of Roof'].replace('Not Applicable', 'Not Applicable - Roof')
fie_df['Type of Walls'] = fie_df['Type of Walls'].replace('Not Applicable', 'Not Applicable - Walls')
fie_df['Tenure Status'] = fie_df['Tenure Status'].replace('Not Applicable', 'Not Applicable - Tenure Status')

In [111]:
from sklearn.preprocessing import OneHotEncoder

In [112]:
fie_df.dtypes

Region                                           object
Total Food Expenditure                            int64
Main Source of Income                            object
Bread and Cereals Expenditure                     int64
Total Rice Expenditure                            int64
Meat Expenditure                                  int64
Total Fish and  marine products Expenditure       int64
Fruit Expenditure                                 int64
Vegetables Expenditure                            int64
Restaurant and hotels Expenditure                 int64
Alcoholic Beverages Expenditure                   int64
Tobacco Expenditure                               int64
Clothing, Footwear and Other Wear Expenditure     int64
Housing and water Expenditure                     int64
Imputed House Rental Value                        int64
Medical Care Expenditure                          int64
Transportation Expenditure                        int64
Communication Expenditure                       

In [113]:
ohe = OneHotEncoder()

In [114]:
object_columns = fie_df.select_dtypes(include=['object']).columns
ordinal_object_columns = ['Household Head Highest Grade Completed', 'Income Class'] 
nominal_object_columns = object_columns[~object_columns.isin(ordinal_object_columns)]
nominal_object_columns

Index(['Region', 'Main Source of Income', 'Household Head Sex',
       'Household Head Marital Status', 'Household Head Occupation',
       'Household Head Class of Worker', 'Type of Household',
       'Type of Building/House', 'Type of Roof', 'Type of Walls',
       'Tenure Status', 'Toilet Facilities', 'Main Source of Water Supply'],
      dtype='object')

In [115]:
feature_array = ohe.fit_transform(fie_df[nominal_object_columns]).toarray()

In [116]:
ohe.categories_

[array(['ARMM', 'CAR', 'I - Ilocos Region', 'II - Cagayan Valley',
        'III - Central Luzon', 'IVA - CALABARZON', 'IVB - MIMAROPA',
        'IX - Zamboanga Peninsula', 'NCR', 'V - Bicol Region',
        'VI - Western Visayas', 'VII - Central Visayas',
        'VIII - Eastern Visayas', 'X - Northern Mindanao',
        'XI - Davao Region', 'XII - SOCCSKSARGEN', 'XIII - Caraga'],
       dtype=object),
 array(['Enterpreneurial Activities', 'Other sources of Income',
        'Wage/Salaries'], dtype=object),
 array(['Female', 'Male'], dtype=object),
 array(['Annulled', 'Divorced/Separated', 'Married', 'Single', 'Unknown',
        'Widowed'], dtype=object),
 array(['Accountants and auditors', 'Accounting and bookkeeping clerks',
        'Administrative secretaries and related associate professionals',
        'Advertising and public relations managers',
        'Agricultural or industrial machinery mechanics and fitters',
        'Agronomists and related scientists',
        'Air traffic 

In [117]:
feature_labels = ohe.categories_

In [118]:
feature_labels = np.array(feature_labels, dtype='object').ravel()
flattened_feature_labels = np.concatenate(feature_labels)

In [119]:
print(flattened_feature_labels)

['ARMM' 'CAR' 'I - Ilocos Region' 'II - Cagayan Valley'
 'III - Central Luzon' 'IVA - CALABARZON' 'IVB - MIMAROPA'
 'IX - Zamboanga Peninsula' 'NCR' 'V - Bicol Region'
 'VI - Western Visayas' 'VII - Central Visayas' 'VIII - Eastern Visayas'
 'X - Northern Mindanao' 'XI - Davao Region' 'XII - SOCCSKSARGEN'
 'XIII - Caraga' 'Enterpreneurial Activities' 'Other sources of Income'
 'Wage/Salaries' 'Female' 'Male' 'Annulled' 'Divorced/Separated' 'Married'
 'Single' 'Unknown' 'Widowed' 'Accountants and auditors'
 'Accounting and bookkeeping clerks'
 'Administrative secretaries and related associate professionals'
 'Advertising and public relations managers'
 'Agricultural or industrial machinery mechanics and fitters'
 'Agronomists and related scientists' 'Air traffic safety technicians'
 'Air transport service supervisors'
 'Aircraft engine mechanics and fitters'
 'Aircraft pilots, navigators and flight engineers'
 'Appraisers and valuers' 'Architects' 'Assembling laborers'
 'Athletes and re

In [120]:
features = pd.DataFrame(feature_array, columns = flattened_feature_labels)

In [121]:
features.head()

ARMM  CAR  I - Ilocos Region  II - Cagayan Valley  III - Central Luzon   
0   0.0  1.0                0.0                  0.0                  0.0  \
1   0.0  1.0                0.0                  0.0                  0.0   
2   0.0  1.0                0.0                  0.0                  0.0   
3   0.0  1.0                0.0                  0.0                  0.0   
4   0.0  1.0                0.0                  0.0                  0.0   

   IVA - CALABARZON  IVB - MIMAROPA  IX - Zamboanga Peninsula  NCR   
0               0.0             0.0                       0.0  0.0  \
1               0.0             0.0                       0.0  0.0   
2               0.0             0.0                       0.0  0.0   
3               0.0             0.0                       0.0  0.0   
4               0.0             0.0                       0.0  0.0   

   V - Bicol Region  ...  Lake, river, rain and others  Others - Water Supply   
0               0.0  ...                           0.0                    0.0  \
1               0.0  ...                           0.0                    0.0   
2               0.0  ...                           0.0                    0.0   
3               0.0  ...                           0.0                    0.0   
4               0.0  ...                           0.0                    0.0   

   Own use, faucet, community water system  Own use, tubed/piped deep well   
0                                      1.0                             0.0  \
1                                      1.0                             0.0   
2                                      0.0                             0.0   
3                                      1.0                             0.0   
4                                      1.0                             0.0   

   Peddler  Protected spring, river, stream, etc   
0      0.0                                   0.0  \
1      0.0                                   0.0   
2      0.0                                   0.0   
3      0.0                                   0.0   
4      0.0                                   0.0   

   Shared, faucet, community water system  Shared, tubed/piped deep well   
0                                     0.0                            0.0  \
1                                     0.0                            0.0   
2                                     1.0                            0.0   
3                                     0.0                            0.0   
4                                     0.0                            0.0   

   Tubed/piped shallow well  Unprotected spring, river, stream, etc  
0                       0.0                                     0.0  
1                       0.0                                     0.0  
2                       0.0                                     0.0  
3                       0.0                                     0.0  
4                       0.0                                     0.0  

[5 rows x 464 columns]

In [122]:
int_df = fie_df.select_dtypes(include=['int64'])
excluded_df = fie_df[ordinal_object_columns]
encoded_fie_df = pd.concat([features, int_df, excluded_df], axis=1)

## Label Encoding of Ordinal Variables

In [123]:
income_class_labels = ["Poor", "Low-income", "Lower middle", "Middle", "Upper middle", "Upper middle but not rich", "Rich"]
highest_grade_completed_labels = ['No Grade Completed', 'Preschool', 'Primary Level', 'Secondary Level', 'Tertiary Level', 'Post Baccalaureate']

income_class_custom_mapping = {label: idx for idx, label in enumerate(income_class_labels)}
highest_grade_completed_custom_mapping = {label: idx for idx, label in enumerate(highest_grade_completed_labels)}

# Transform your data using the custom mapping
encoded_fie_df[ordinal_object_columns[0]] = fie_df[ordinal_object_columns[0]].map(highest_grade_completed_custom_mapping)
encoded_fie_df[ordinal_object_columns[1]] = fie_df[ordinal_object_columns[1]].map(income_class_custom_mapping)

# Feature Selection

In [217]:
from sklearn.model_selection import cross_val_predict, cross_validate
from sklearn.metrics import classification_report, confusion_matrix
import seaborn as sns
from sklearn.neighbors import KNeighborsClassifier
from sklearn.feature_selection import SequentialFeatureSelector
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.neural_network import MLPClassifier 
from sklearn import tree

In [125]:
encoded_fie_df

ARMM  CAR  I - Ilocos Region  II - Cagayan Valley  III - Central Luzon   
0       0.0  1.0                0.0                  0.0                  0.0  \
1       0.0  1.0                0.0                  0.0                  0.0   
2       0.0  1.0                0.0                  0.0                  0.0   
3       0.0  1.0                0.0                  0.0                  0.0   
4       0.0  1.0                0.0                  0.0                  0.0   
...     ...  ...                ...                  ...                  ...   
41539   0.0  0.0                0.0                  0.0                  0.0   
41540   0.0  0.0                0.0                  0.0                  0.0   
41541   0.0  0.0                0.0                  0.0                  0.0   
41542   0.0  0.0                0.0                  0.0                  0.0   
41543   0.0  0.0                0.0                  0.0                  0.0   

       IVA - CALABARZON  IVB - MIMAROPA  IX - Zamboanga Peninsula  NCR   
0                   0.0             0.0                       0.0  0.0  \
1                   0.0             0.0                       0.0  0.0   
2                   0.0             0.0                       0.0  0.0   
3                   0.0             0.0                       0.0  0.0   
4                   0.0             0.0                       0.0  0.0   
...                 ...             ...                       ...  ...   
41539               0.0             0.0                       0.0  0.0   
41540               0.0             0.0                       0.0  0.0   
41541               0.0             0.0                       0.0  0.0   
41542               0.0             0.0                       0.0  0.0   
41543               0.0             0.0                       0.0  0.0   

       V - Bicol Region  ...  Number of Airconditioner   
0                   0.0  ...                         0  \
1                   0.0  ...                         0   
2                   0.0  ...                         0   
3                   0.0  ...                         0   
4                   0.0  ...                         0   
...                 ...  ...                       ...   
41539               0.0  ...                         0   
41540               0.0  ...                         0   
41541               0.0  ...                         0   
41542               0.0  ...                         0   
41543               0.0  ...                         0   

       Number of Car, Jeep, Van  Number of Landline/wireless telephones   
0                             0                                       0  \
1                             0                                       0   
2                             0                                       0   
3                             0                                       0   
4                             0                                       0   
...                         ...                                     ...   
41539                         0                                       0   
41540                         0                                       0   
41541                         0                                       0   
41542                         0                                       0   
41543                         0                                       0   

       Number of Cellular phone  Number of Personal Computer   
0                             2                            1  \
1                             3                            1   
2                             0                            0   
3                             1                            0   
4                             3                            0   
...                         ...                          ...   
41539                         1                            0   
41540                         3    

In [126]:
X = encoded_fie_df.iloc[:, :-1]
y = encoded_fie_df.iloc[:, -1:].values.ravel()

## 1. Complete Features - Hannah

### A. KNN

In [127]:
model = KNeighborsClassifier(n_neighbors=15)
y_pred = cross_val_predict(estimator=model, X=X, y=y, cv=10)
print(classification_report(y, y_pred))

              precision    recall  f1-score   support

           0       0.72      0.72      0.72      3282
           1       0.70      0.81      0.75     12299
           2       0.70      0.72      0.71     13989
           3       0.62      0.57      0.59      7077
           4       0.60      0.48      0.53      3389
           5       0.58      0.31      0.41      1108
           6       0.85      0.38      0.53       400

    accuracy                           0.68     41544
   macro avg       0.68      0.57      0.61     41544
weighted avg       0.68      0.68      0.68     41544



### B. SVM

In [ ]:
model = svm.SVC()
y_pred = cross_val_predict(estimator=model, X=X, y=y, cv=10)
print(classification_report(y, y_pred))

### C. MLP

In [ ]:
model = MLPClassifier(random_state=1, max_iter=300)
y_pred = cross_val_predict(estimator=model, X=X, y=y, cv=10)
print(classification_report(y, y_pred))

### D. DT

In [ ]:
model = tree.DecisionTreeClassifier()
y_pred = cross_val_predict(estimator=model, X=X, y=y, cv=10)
print(classification_report(y, y_pred))

## 2. PCA Components as New Features - Hannah

In [ ]:
from sklearn.preprocessing import RobustScaler
from sklearn.decomposition import PCA

scaler = RobustScaler()
X_scaled = scaler.fit_transform(X)
X_scaled

In [ ]:
pca = PCA(0.95)
X_pca_components = pca.fit_transform(X_scaled)
X_pca_components.shape

### A. KNN

In [ ]:
model = KNeighborsClassifier(n_neighbors=15)
y_pred = cross_val_predict(estimator=model, X=X_pca_components, y=y, cv=10)
print(classification_report(y, y_pred))

### B. SVM

In [ ]:
model = svm.SVC()
y_pred = cross_val_predict(estimator=model, X=X_pca_components, y=y, cv=10)
print(classification_report(y, y_pred))

### C. MLP

In [ ]:
model = MLPClassifier(random_state=1, max_iter=300)
y_pred = cross_val_predict(estimator=model, X=X_pca_components, y=y, cv=10)
print(classification_report(y, y_pred))

### D. DT

In [ ]:
model = tree.DecisionTreeClassifier()
y_pred = cross_val_predict(estimator=model, X=X_pca_components, y=y, cv=10)
print(classification_report(y, y_pred))

## 3. Features selected from PCA Components - Hannah

In [ ]:
from sklearn.preprocessing import RobustScaler
from sklearn.decomposition import PCA

scaler = RobustScaler()
X_scaled = scaler.fit_transform(X)
X_scaled

In [ ]:
pca = PCA(0.95)
X_pca = pca.fit_transform(X_scaled)
X_pca.shape

In [ ]:
cumulative_sum = np.cumsum(pca.explained_variance_ratio_)
cumulative_sum[-1]

In [ ]:
components = pca.components_
num_components = components.shape[0]
highest_loading_indices = set()

In [ ]:
for i in range(num_components):
    loading_scores = np.abs(components[i, :])
    highest_loading_indices.add(np.argmax(loading_scores))

In [ ]:
highest_loading_indices = list(highest_loading_indices)
len(highest_loading_indices)

In [ ]:
X_pca_features = X.iloc[:,highest_loading_indices]
X_pca_features.shape

In [ ]:
X_scaled

In [ ]:
X_pca_features

### A. KNN

In [ ]:
model = KNeighborsClassifier(n_neighbors=15)
y_pred = cross_val_predict(estimator=model, X=X_pca_features, y=y, cv=10)
print(classification_report(y, y_pred))

### B. SVM

In [ ]:
model = svm.SVC()
y_pred = cross_val_predict(estimator=model, X=X_pca_features, y=y, cv=10)
print(classification_report(y, y_pred))

### C. MLP

In [ ]:
model = MLPClassifier(random_state=1, max_iter=300)
y_pred = cross_val_predict(estimator=model, X=X_pca_features, y=y, cv=10)
print(classification_report(y, y_pred))

### D. DT

In [ ]:
model = tree.DecisionTreeClassifier()
y_pred = cross_val_predict(estimator=model, X=X_pca_features, y=y, cv=10)
print(classification_report(y, y_pred))

## 4. Forward Selection

### A. KNN

In [218]:
model = KNeighborsClassifier(n_neighbors=15)
sfs = SequentialFeatureSelector(model)
sfs.fit(X, y)
X_fs_knn = sfs.transform(X)
X_fs_knn

KeyboardInterrupt: 

## 5. Decision Tree

In [129]:
print(encoded_fie_df.shape)
print(X.shape)
print(y.shape)

(41544, 509)
(41544, 508)
(41544,)


In [ ]:
tree_mod = tree.DecisionTreeClassifier()
output = cross_validate(estimator=tree_mod, X=X, y=y, cv=10, return_estimator=True)

In [145]:
feature_importances = np.zeros(X.shape[1])
for dtc in output['estimator']:
    feature_importances += dtc.feature_importances_

feature_importances = feature_importances / 10


In [210]:
dt_features = []
argsort_fi = np.argsort(feature_importances)

max_idx = feature_importances.shape[0]
min_idx = max_idx - int(np.floor(feature_importances.shape[0] * 0.05))

final_features = argsort_fi[min_idx:max_idx]

for f in final_features:
    dt_features.append(X.columns[f])

X_dt = X[dt_features]

In [220]:
dt_features

['Number of bedrooms',
 'Tobacco Expenditure',
 'Imputed House Rental Value',
 'Alcoholic Beverages Expenditure',
 'Bread and Cereals Expenditure',
 'Total number of family members employed',
 'House Age',
 'Total Rice Expenditure',
 'Total Fish and  marine products Expenditure',
 'Restaurant and hotels Expenditure',
 'House Floor Area',
 'Vegetables Expenditure',
 'Fruit Expenditure',
 'Household Head Age',
 'Education Expenditure',
 'Meat Expenditure',
 'Special Occasions Expenditure',
 'Clothing, Footwear and Other Wear Expenditure',
 'Medical Care Expenditure',
 'Communication Expenditure',
 'Transportation Expenditure',
 'Total Income from Entrepreneurial Acitivites',
 'Miscellaneous Goods and Services Expenditure',
 'Housing and water Expenditure',
 'Total Food Expenditure']

### A. KNN

In [211]:
model = KNeighborsClassifier(n_neighbors=15)
y_pred = cross_val_predict(estimator=model, X=X_dt, y=y, cv=10)
print(classification_report(y, y_pred))

              precision    recall  f1-score   support

           0       0.73      0.72      0.73      3282
           1       0.71      0.81      0.76     12299
           2       0.70      0.72      0.71     13989
           3       0.62      0.57      0.60      7077
           4       0.60      0.48      0.53      3389
           5       0.59      0.32      0.42      1108
           6       0.85      0.40      0.54       400

    accuracy                           0.69     41544
   macro avg       0.69      0.58      0.61     41544
weighted avg       0.68      0.69      0.68     41544



### B. SVM

In [212]:
model = svm.SVC()
y_pred = cross_val_predict(estimator=model, X=X_dt, y=y, cv=10)
print(classification_report(y, y_pred))

              precision    recall  f1-score   support

           0       0.79      0.66      0.72      3282
           1       0.74      0.81      0.77     12299
           2       0.74      0.76      0.75     13989
           3       0.67      0.66      0.66      7077
           4       0.65      0.56      0.60      3389
           5       0.63      0.42      0.50      1108
           6       0.78      0.51      0.62       400

    accuracy                           0.72     41544
   macro avg       0.71      0.62      0.66     41544
weighted avg       0.72      0.72      0.72     41544



### C. MLP

In [213]:
model = MLPClassifier(random_state=1, max_iter=300)
y_pred = cross_val_predict(estimator=model, X=X_dt, y=y, cv=10)
print(classification_report(y, y_pred))

              precision    recall  f1-score   support

           0       0.62      0.45      0.52      3282
           1       0.51      0.63      0.56     12299
           2       0.50      0.31      0.38     13989
           3       0.35      0.42      0.38      7077
           4       0.29      0.41      0.34      3389
           5       0.23      0.23      0.23      1108
           6       0.20      0.38      0.26       400

    accuracy                           0.44     41544
   macro avg       0.39      0.40      0.38     41544
weighted avg       0.46      0.44      0.44     41544



### D. DT

In [214]:
model = tree.DecisionTreeClassifier()
y_pred = cross_val_predict(estimator=model, X=X_dt, y=y, cv=10)
print(classification_report(y, y_pred))

              precision    recall  f1-score   support

           0       0.62      0.63      0.62      3282
           1       0.66      0.66      0.66     12299
           2       0.63      0.64      0.63     13989
           3       0.53      0.53      0.53      7077
           4       0.47      0.48      0.48      3389
           5       0.40      0.37      0.38      1108
           6       0.50      0.50      0.50       400

    accuracy                           0.60     41544
   macro avg       0.55      0.54      0.54     41544
weighted avg       0.60      0.60      0.60     41544



# Data Visualization - Hannah

K Means + SOM

## K Means

In [ ]:
from sklearn.cluster import KMeans
from sklearn.metrics import davies_bouldin_score

In [ ]:
inertias = []
dbi_scores = []
mapping = {}
K = range(2, 21)

for k in K:
    kMeans = KMeans(n_clusters=k, random_state=0, n_init=10)
    kMeans.fit(X)
    inertias.append(kMeans.inertia_)
    mapping[k] = kMeans.inertia_
    
    dbi = davies_bouldin_score(X, kMeans.labels_)
    dbi_scores.append(dbi)
    #todo: print dbi

In [ ]:
dbi_scores

In [ ]:
plt.plot(K, inertias, 'bx-')
plt.xlabel('Values of K')
plt.ylabel('Inertia')
plt.title('The Elbow Method Using Inertia')
plt.xlim([2, 20])
plt.show()

## SOM

In [ ]:
from minisom import MiniSom
from sklearn.cluster import KMeans

In [ ]:
data = X.to_numpy()
som_shape = (10,10)
som = MiniSom(som_shape[0], som_shape[1], data.shape[1])
som.train_batch(data, 100000)

centroids = som.get_weights()
centroids.shape

In [ ]:
flattened_centroids = centroids.reshape((centroids.shape[0]*centroids.shape[1], centroids.shape[2]))
flattened_centroids.shape

### 2. Weights of each SOM node are the input to k-means clustering

In [ ]:
centroids_df = pd.DataFrame(flattened_centroids, columns=X.columns)

In [ ]:
kMeans = KMeans(n_clusters=7, random_state=0, n_init=10)
kMeans.fit(centroids_df)
centroids_df['cluster'] = kMeans.labels_
centroids_df

## 3. Label the SOM nodes

In [ ]:
from sklearn.neighbors import NearestNeighbors

In [ ]:
centroids_features = centroids_df.drop('cluster', axis=1).values

In [ ]:
X_features = X.values

In [ ]:
k_neighbors = 7
nn_model = NearestNeighbors(n_neighbors=k_neighbors, metric='euclidean')
nn_model.fit(X_features)
distances, indices = nn_model.kneighbors(centroids_features)

In [ ]:
majority_class = []
for i in indices:
    majority_class.append(fie_df['Income Class'].values[i].max(axis=0))
centroids_df['majority class'] = majority_class

In [ ]:
centroids_df

## 4. Present the SOM where nodes are labeled and clustered.

In [ ]:
clusters = centroids_df['cluster'].to_numpy()
clusters = clusters.reshape((centroids.shape[0], centroids.shape[1]))
clusters.shape

In [ ]:
majority_class_labels = centroids_df['majority class'].to_numpy()
majority_class_labels = majority_class_labels.reshape((centroids.shape[0], centroids.shape[1]))
majority_class_labels.shape

In [ ]:
fig, ax = plt.subplots(figsize=(16, 16))
heatmap = ax.pcolormesh(clusters, edgecolors='w', linewidth=0.5)

for i in range(len(clusters)):
    for j in range(len(clusters[i])):
        ax.text(j + 0.5, i + 0.5, str(majority_class_labels[i, j]), ha='center', va='center', color='black', fontsize=7)
plt.tight_layout()
plt.show()

# References
Domingo, K. (2020 September 17). Who are identified rich, poot? Gov't shows income class brackets in PH. ABSCBN News. https://news.abs-cbn.com/news/09/17/20/who-are-identified-rich-poor-govt-shows-income-class-brackets-in-ph
 
Philippine Statistics Authority. (2017). 2015 Family income and expenditure survey. https://library.psa.gov.ph/cgi-bin/koha/opac-detail.pl?biblionumber=15585 

# DRAFT

## Train-Test Split

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X = encoded_fie_df.iloc[:, :-1]
y = encoded_fie_df.iloc[:, -1:]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=123)